test/lit/passes/gufa-tnh.wast

```diff
@@ -0,0 +1,1905 @@
+
+  ;; CHECK:      (type $anyref_=>_none (func (param anyref)))
+
+  ;; CHECK:      (export "out" (func $caller))
+
+  ;; CHECK:      (func $called (type $ref?|$A|_ref?|$A|_ref?|$B|_ref?|$B|_ref?|$B|_ref?|$B|_ref?|$B|_ref?|$B|_ref?|$A|_=>_none) (param $struct.get (ref null $A)) (param $struct.set (ref null $A)) (param $array.get (ref null $B)) (param $array.set (ref null $B)) (param $array.len (ref null $B)) (param $array.copy.src (ref null $B)) (param $array.copy.dest (ref null $B)) (param $array.fill (ref null $B)) (param $ref.test (ref null $A))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 0)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (struct.set $A 0
+  ;; CHECK-NEXT:   (local.get $struct.set)
+  ;; CHECK-NEXT:   (i32.const 0)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (array.get $B
+  ;; CHECK-NEXT:    (local.get $array.get)
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (array.set $B
+  ;; CHECK-NEXT:   (local.get $array.set)
+  ;; CHECK-NEXT:   (i32.const 2)
+  ;; CHECK-NEXT:   (i32.const 3)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (array.len
+  ;; CHECK-NEXT:    (local.get $array.len)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (array.copy $B $B
+  ;; CHECK-NEXT:   (local.get $array.copy.src)
+  ;; CHECK-NEXT:   (i32.const 4)
+  ;; CHECK-NEXT:   (local.get $array.copy.dest)
+  ;; CHECK-NEXT:   (i32.const 5)
+  ;; CHECK-NEXT:   (i32.const 6)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (array.fill $B
+  ;; CHECK-NEXT:   (local.get $array.fill)
+  ;; CHECK-NEXT:   (i32.const 7)
+  ;; CHECK-NEXT:   (i32.const 8)
+  ;; CHECK-NEXT:   (i32.const 9)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.test $A
+  ;; CHECK-NEXT:    (local.get $ref.test)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $called
+    (param $struct.get (ref null $A))
+    (param $struct.set (ref null $A))
+    (param $array.get (ref null $B))
+    (param $array.set (ref null $B))
+    (param $array.len (ref null $B))
+    (param $array.copy.src (ref null $B))
+    (param $array.copy.dest (ref null $B))
+    (param $array.fill (ref null $B))
+    (param $ref.test (ref null $A))
+
+    ;; All the operations trap on null, aside from ref.test.
+    (drop
+      (struct.get $A 0
+        (local.get $struct.get)
+      )
+    )
+    (struct.set $A 0
+      (local.get $struct.set)
+      (i32.const 0)
+    )
+    (drop
+      (array.get $B
+        (local.get $array.get)
+        (i32.const 1)
+      )
+    )
+    (array.set $B
+      (local.get $array.set)
+      (i32.const 2)
+      (i32.const 3)
+    )
+    (drop
+      (array.len $B
+        (local.get $array.len)
+      )
+    )
+    (array.copy $B $B
+      (local.get $array.copy.src)
+      (i32.const 4)
+      (local.get $array.copy.dest)
+      (i32.const 5)
+      (i32.const 6)
+    )
+    (array.fill $B
```

Comment:
Do we want `array.init_data` and `array.init_elem` as well?

Reply:
Oh, for some reason I thought those create new arrays... yeah, I'll add those.

Reply:
You were probably thinking of `array.new_data` and `arrray.new_elem` :)